langchain/vectorstores/pgvector.py

```diff
@@ -0,0 +1,421 @@
+        "Please install it with `pip install sqlalchemy`."
+    )
+
+from langchain.docstore.document import Document
+from langchain.embeddings.base import Embeddings
+from langchain.utils import get_from_dict_or_env
+from langchain.vectorstores.base import VectorStore
+
+try:
+    from pgvector.sqlalchemy import Vector
+except ImportError:
+    raise ValueError(
+        "Could not import pgvector python package. "
+        "Please install it with `pip install pgvector`."
+    )
+
+
+Base = declarative_base()  # type: Any
+
+
+ADA_TOKEN_COUNT = 1536
+_LANGCHAIN_DEFAULT_COLLECTION_NAME = "langchain"
+
+
+class BaseModel(Base):
+    __abstract__ = True
+    uuid = sqlalchemy.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
+
+
+class CollectionStore(BaseModel):
+    __tablename__ = "langchain_pg_collection"
+
+    name = sqlalchemy.Column(sqlalchemy.String)
+    cmetadata = sqlalchemy.Column(sqlalchemy.JSON)
+
+    embeddings = relationship(
+        "EmbeddingStore",
+        back_populates="collection",
+        passive_deletes=True,
+    )
+
+    @classmethod
+    def get_by_name(cls, session: Session, name: str) -> Optional["CollectionStore"]:
+        return session.query(cls).filter(cls.name == name).first()
+
+    @classmethod
+    def get_or_create(
+        cls,
+        session: Session,
+        name: str,
+        cmetadata: Optional[dict] = None,
+    ) -> Tuple["CollectionStore", bool]:
+        """
+        Get or create a collection.
+        Returns [Collection, bool] where the bool is True if the collection was created.
+        """
+        created = False
+        collection = cls.get_by_name(session, name)
+        if collection:
+            return collection, created
+
+        collection = cls(name=name, metadata=cmetadata)
+        session.add(collection)
+        session.commit()
+        created = True
+        return collection, created
+
+
+class EmbeddingStore(BaseModel):
+    __tablename__ = "langchain_pg_embedding"
+
+    collection_id: Mapped[UUID] = sqlalchemy.Column(
+        UUID(as_uuid=True),
+        sqlalchemy.ForeignKey(
+            f"{CollectionStore.__tablename__}.uuid",
+            ondelete="CASCADE",
+        ),
+    )
+    collection = relationship(CollectionStore, back_populates="embeddings")
+
+    embedding: Vector = sqlalchemy.Column(Vector(ADA_TOKEN_COUNT))
+    document = sqlalchemy.Column(sqlalchemy.String, nullable=True)
+    cmetadata = sqlalchemy.Column(sqlalchemy.JSON, nullable=True)
+
+    # custom_id : any user defined id
+    custom_id = sqlalchemy.Column(sqlalchemy.String, nullable=True)
+
+
+class QueryResult:
+    EmbeddingStore: EmbeddingStore
+    distance: float
+
+
+class DistanceStrategy(str, enum.Enum):
+    EUCLIDEAN = EmbeddingStore.embedding.l2_distance
+    COSINE = EmbeddingStore.embedding.cosine_distance
+    MAX_INNER_PRODUCT = EmbeddingStore.embedding.max_inner_product
+
+
+DEFAULT_DISTANCE_STRATEGY = DistanceStrategy.EUCLIDEAN
+
+
+@dataclass
```

Comment:
we use pydantic BaseModel rather than dataclass

Reply:
I mainly wanted to use the __post_init__ functionality from the dataclass to create the necessary tables and collection, but if that's the case, I'll change it to normal class.